### ISSUE TITLE
Cut benchmark suite down to ~1 hour and run it with each pull request

### ISSUE THREAD

**Issue by mjambon** [semgrep/semgrep#3882]
Right now we have a daily job that runs lengthy benchmarks, but those benchmarks fail and we don't notice them or fix them, and they're not usable when we need them. We should have usable benchmarks, that remain usable.

Proposal:

* reduce full suite so that it takes about 1 hour and keep it that way (by force-aborting the whole thing if it takes more than e.g. 90 min)
* keep running quick suite suite on each PR
* run full suite on each PR as well but don't block the merge (so that we'll notice, at least sometimes, if it crashes)
* keep daily runs of the full suite on the main branch (which now also runs on each PR) so as to post 1 data point per day, producing nice graphs

**Comment by mjambon** [comment]
Possible implementation with CircleCI:
* use CircleCI to run the comparative benchmarks (full suite, baseline vs. new branch)
* trigger the job conditionally if the branch name is of the form `*-bench`
* optionally* post a summary of the benchmark results as a PR comment

*Posting PR comments from CircleCI is possible but seems to require a GitHub token with full write permissions, which isn't great.

**Comment by IagoAbal** [comment]
Did you try running the full benchmarks? How long did they take to run for you? There is a comment that claims that `--all` will take more than a day to run, but Semgrep has gotten much faster, plus I guess we only want to run `--std-only`.

**Comment by r2c-demo** [comment]
This issue is synced in Linear at https://linear.app/r2c/issue/PA-354/cut-benchmark-suite-down-to-1-hour-and-run-it-with-each-pull-request.

**Comment by IagoAbal** [comment]
The current benchmark suite takes ~3 hours on my laptop which I think it's totally fine, even for CI. I would say we need to expand it rather than cutting it down! We should run the entire registry on each project IMO, and use this to collect stats.

**Comment by mjambon** [comment]
1 hour is about how much patience I have to wait for test results. That's why I set the limit to 1 hour. Often I can arrange to take a lunch break or exercise during that time. I can't do that if the job takes 2 hours. Then it forces me to jump to another task and lose context.

With a bit of effort in selecting benchmarks judiciously, I don't see why we couldn't reduce this time arbitrarily. Just 10 minutes is already a really long time, given that the median semgrep scan takes 45 seconds. If you think a particular benchmark is valuable and necessarily must take over 10 minutes, I'd like to hear why.

**Comment by mjambon** [comment]
The issue with using GHA to trigger a workflow is that it's a whole workflow, not just an optional job. This would require duplicating the workflow config.

**Comment by IagoAbal** [comment]
I don't understand the issue with the time here. We have a quick benchmark suite that is blocking (you can't merge if it doesn't pass). But this one should not be blocking, right? You should be fine to merge even if the long benchmarks have not completed. And, if they fail, then you'll get some kind of notification and you can look into it later on. So you may want it to be more complete rather than faster.

**Comment by emjin** [comment]
It's nice to have the green checkmark before merging imo

**Comment by emjin** [comment]
We should really be running the benchmarks in parallel though; that would probably allow us to run way more

**Comment by mjambon** [comment]
> You should be fine to merge even if the long benchmarks have not completed.

That's what we'd like to avoid. The scenario is:
1. I know that I touched some performance-sensitive code.
2. I'd like to be sure that everything is still fine performance-wise.
3. I'm ok with waiting an hour or so before merging.

Another scenario is the quick-benchmarks:
1. It's a regular pull request and I don't expect any performance issues.
2. The quick benchmarks are a sanity check, in case I did something really bad without realizing it.

We really only need to block the merge for scenarios like the second one, where nothing bad is expected.

**Comment by mjambon** [comment]
> We should really be running the benchmarks in parallel though; that would probably allow us to run way more

An issue is to ensure our own jobs don't compete for resources on the same machine.

**Comment by emjin** [comment]
CircleCI has this: https://circleci.com/docs/2.0/parallelism-faster-jobs/. Dropping it here; will investigate
